daemon/qrexec-client.c

```diff
@@ -51,6 +52,57 @@ int child_exited = 0;
 
 extern char **environ;
 
+static char *xstrdup(const char *arg) {
+    char *retval = strdup(arg);
+    if (!retval) {
+        fputs("Out of memory in xstrdup()\n", stderr);
+        abort();
+    }
+    return retval;
+}
+
+static void set_remote_domain(const char *src_domain_name) {
+    if (setenv("QREXEC_REMOTE_DOMAIN", src_domain_name, 1)) {
+        fputs("Cannot set QREXEC_REMOTE_DOMAIN\n", stderr);
+        abort();
+    }
+}
+
+static void close_stdin_fd(void) {
+    if (local_stdin_fd < 0)
+        return;
+    if (shutdown(local_stdin_fd, SHUT_WR)) {
+        fputs("Cannot shutdown socket\n", stderr);
+        abort();
```

Comment:
I think now it breaks non-socket case. On pipes shutdown will fail with ENOTSOCK and it shouldn't be fatal.
See how it's handled in the qrexec-agent side.

Reply:
Fixed by not considering ENOTSOCK fatal.